### ISSUE TITLE
invalid parameter names in EC2

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaCloud/AWS.jl#444]
I tired running `Ec2.run_instances` and got the following error from AWS:
```
<Errors><Error><Code>UnknownParameter</Code><Message>The parameter clientToken is not recognized</Message></Error></Errors>
```

Seems that the issue is that this parameter is out of date.  Looking at the [documentation](https://juliacloud.github.io/AWS.jl/stable/services/ec2.html#Main.Ec2.run_instances-Tuple{Any,%20Any}) confirms that AWS.jl is using a field called 'clientToken' but the [AWS docs](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html) call this `ClientToken`.

Not sure if this is just a routine API building update or there is something else wrong?

**Comment by mattBrzezinski** [comment]
Taken from [aws-sdk-js](https://raw.githubusercontent.com/aws/aws-sdk-js/master/apis/ec2-2016-11-15.normal.json)

```
        "ClientToken": {
          "shape": "String",
          "documentation": "<p>Unique, case-sensitive identifier you provide to ensure the idempotency of the request. If you do not specify a client token, a randomly generated token is used for the request to ensure idempotency.</p> <p>For more information, see <a href=\"https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html\">Ensuring Idempotency</a>.</p> <p>Constraints: Maximum 64 ASCII characters</p>",
          "idempotencyToken": true,
          "locationName": "clientToken"
        },
```

The name *should* be `clientToken`. And no additional locations (such as this needs to be in the headers) is specified.

**Comment by mattBrzezinski** [comment]
Could you try the following? This will submit the request w/o a `clientToken`. According to the docs AWS will just set one for you.

```julia
using AWS
using AWS.AWSServices: ec2

ec2("RunInstances", Dict("MaxCount"=>1, "MinCount"=>2))
```

**Comment by ExpandingMan** [comment]
I don't understand at all what's going on here, but the code you posted *does* work.

**Comment by mattBrzezinski** [comment]
> I don't understand at all what's going on here, but the code you posted _does_ work.

Interesting... might be a problem with how we generate stuff. Or a problem w/ Amazon themselves.

---

The `AWS.jl` package basically has two definitions for making requests. I've named these `high-level` and `low-level` code. The `low-level` stuff is what I posted above. It contains definitions for each service, including how requests are made, e.g. EC2 is a [QueryService](https://github.com/JuliaCloud/AWS.jl/blob/9a37112f63cc38a59f6b255bb1aff74fc697eec5/src/AWSServices.jl#L203).

You can make requests using this low-level code, but it's much more annoying because you need to know more details about the request itself. Such as, what it's called in the API, what are required parameters, where optional parameters are placed (body? headers?).

It's stuff that the general person shouldn't care about. This is where the `high-level` code comes in and is probably what you're more familiar with. Just do;

```julia
using AWS
@service EC2

?EC2.run_instances
```

And you can see what the required params are when calling the Julia function. And how to place the optional ones as well.

**Comment by ExpandingMan** [comment]
Ah makes sense.  So assuming you made the request you showed above today, it would seem that AWS's SDK is returning deprecated values.  Not sure what, if anything, can be done about this...

**Comment by mattBrzezinski** [comment]
> Ah makes sense. So assuming you made the request you showed above today, it would seem that AWS's SDK is returning deprecated values. Not sure what, if anything, can be done about this...

To me it seems that either:

1. The optional parameter does not exist anymore and things were not properly updated on AWS part
2. The parameter does exist, but needs more information, such as being in a header

Might be worthwhile to look into other endpoints which use idempotent parameters and see if they come across the same problem. Would you mind trying to run the code below and seeing if it works? Might just be that the parameter AWS expects actually is with a capital `C`.

```julia
using AWS.AWSServices: ec2

ec2("RunInstances", Dict("MaxCount"=>2, "MinCount"=>1), "ClientToken"=>"foobar") 
```

**Comment by ExpandingMan** [comment]
Was back at this today... so `"ClientToken"` does indeed work... from what I can tell, the AWS documentation seems to be indicating that the first letter is capitalized for requests but not responses, which is why both versions appear so much in the documentation.  This is definitely an error on the wrapper side, but whether that's caused by an error in whatever you're using to generate all the calls I still don't know.

**Comment by mattBrzezinski** [comment]
I just cut the issue above w/ AWS. Either I am misunderstanding something with `locationName` in their definitions or there's a bug. Hopefully we get a response back soon.

I'd say for now, just do a "low-level" request to work around the problem.

**Comment by ExpandingMan** [comment]
Yeah, I can just do the lower level `ec2`, it's very easy to work around.

**Comment by ExpandingMan** [comment]
Just FYI, this also just happened to me with `instanceId` in `ModifyInstanceAttribute`.  Some thing, AWS expects `InstanceId`, it gets `instanceId`.

**Comment by ExpandingMan** [comment]
Much to my surprise I'm still getting burned by this.

**Comment by mattBrzezinski** [comment]
> Much to my surprise I'm still getting burned by this.

I'll follow up on this, we reached out to our AWS rep late last year, haven't heard anything back.

**Comment by ExpandingMan** [comment]
It looks like a wrong name is getting generated somewhere (`clientToken` rather than `ClientToken`) but the code generation part of this package is still rather mysterious to me, so I haven't even looked into how it would be fixed.

**Comment by mattBrzezinski** [comment]
> It looks like a wrong name is getting generated somewhere (`clientToken` rather than `ClientToken`) but the code 
generation part of this package is still rather mysterious to me, so I haven't even looked into how it would be fixed.


Here's a link to the issue cut w/ the SDK that we generate our code off of,
* https://github.com/aws/aws-sdk-js/issues/3892

Stuff like botocore special cases stuff like this,
* https://github.com/boto/botocore/blob/c920c8f40c6fb61b9b3f3b1d76ff3c14fdad3490/botocore/serialize.py#L319

Using the `ClientToken` example from the above issue,

```json
"ClientToken": {
  "shape": "String",
  "documentation": "<p>Unique, case-sensitive identifier you provide to ensure the idempotency of the request. If you do not specify a client token, a randomly generated token is used for the request to ensure idempotency.</p> <p>For more information, see <a href=\"https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html\">Ensuring Idempotency</a>.</p> <p>Constraints: Maximum 64 ASCII characters</p>",
  "idempotencyToken": true,
  "locationName": "clientToken"
},
```

From the blob above it seems like `clientToken` is what should be expected, but that's not correct. It wants `ClientToken`. I need to look into this again, but from memory I think this `locationName` usage is also inconsistent, so we can't even say when generating EC2 Julia definitions just ignore `locationName`.

AWS is confusing and hard.

**Comment by ExpandingMan** [comment]
Yeah, their API is horrendous.

I don't suppose there's any chance that it's as simple as it always needing to be the key of that dict rather than `locationName`?

I vaguely remember there being some different convention for inputs and outputs... like maybe the input value was `ClientToken` but if it ever gets returned it's `clientToken`... something like that.

**Comment by mattBrzezinski** [comment]
> Yeah, their API is horrendous.
> 
> I don't suppose there's any chance that it's as simple as it always needing to be the key of that dict rather than `locationName`?
> 
> I vaguely remember there being some different convention for inputs and outputs... like maybe the input value was `ClientToken` but if it ever gets returned it's `clientToken`... something like that.

This is where it gets interesting... **usually** if the `locationName` is present, that's what AWS wants. If it DNE, then you use the top level key. EC2 doesn't follow that convention seemingly though 🙃

**Comment by iamed2** [comment]
The botocore link above clarifies the issue: the `ec2` protocol requires the capitalization. The service spec for EC2 claims to use the `ec2` protocol, not the `query` protocol. The issue with AWS.jl is that it doesn't implement any different behaviour for the `ec2` protocol, it assumes that `ec2` protocol is the same as `query` protocol. 

As far as we know right now, "implementing the EC2 protocol" is just putting in place the capitalization that botocore does when `protocol == "ec2"`. I think that means threading the protocol through `_get_function_parameters` to make sure `locationName` is capitalized.
